fix: sync Zustand store with SSR props after client navigation

Harshit-Mourya opened the pull request:
This PR fixes an issue where the Zustand store (`pageName`) would not reflect the correct SSR value during client-side navigation in a Next.js app.

### Fix Details:
- Introduced a custom `useSyncPageName` hook that syncs the `pageName` prop from `getServerSideProps` into the Zustand store on mount.
- Replaced `uiStore.js` with a fully typed `uiStore.ts`.
- Ensured the hydration is done safely and only when needed.
- This approach avoids stale client state after navigation without triggering extra re-renders.

Fixes #1943

Harshit-Mourya committed:
fix: sync Zustand store with SSR props after client navigation

Harshit-Mourya committed:
Refactor: use per-request Zustand store with createUiStore

Harshit-Mourya commented:
Update: Added per-request store using `createUiStore()` to fix hydration issues on client navigation. Refactored `_app.tsx` and created context-based `UiStoreContext`.

andrey-sheremeta commented on src/pages/_app.tsx:
```diff
@@ -1,19 +1,28 @@
 export default function App({ Component, pageProps }: AppProps) {
-    const store = useInitUiStore(pageProps)
+  const storeRef = useRef(
+    createUiStore({
```
it will impact the performance of the application, both server & client. We can't create a new instance each time, for each request.

andrey-sheremeta commented on src/hooks/useSyncPageName.ts:
```diff
@@ -0,0 +1,12 @@
+export const useSyncPageName = (pageName: string) => {
+  const store = useContext(UiStoreContext);
+
+  useEffect(() => {
```
It should be implemented in a way that allows us to update all properties in one place. Imagine if you have 10 props, we won't create 10 hooks, right? It should be one hook that will handle all props.

andrey-sheremeta replied:
But in general, it doesn't solve the initial problem. On page load, we first receive the old value and only after the update in "useEffect" we receive a fresh one.

Harshit-Mourya replied:
Thanks for the feedback!
I implemented it that way initially based on your earlier suggestion to create a new store per request to avoid shared state during SSR.

But I now understand the concern regarding performance when recreating the store on each render.
I’ll update the logic to ensure a new store is created per request only on the server side, and reused on the client.

I’ll also refactor the useSyncPageName hook into a general one that handles syncing all props together.

Harshit-Mourya committed:
fix: sync Zustand store with SSR props and prevent stale state on client navigation

Harshit-Mourya commented:
@andrey-sheremeta 

Thanks a lot for the detailed feedback!
I've updated the implementation based on your suggestions:
The Zustand store is now created per request only on the server side, and reused on the client to improve performance.
Introduced a generalized useSyncUiProps hook that syncs all SSR props into the store at once.
Ensured hydration is handled cleanly to avoid stale state on initial load.

Let me know if you'd like any further adjustments!
